Xamarin.Forms.Core.UnitTests/NavigationUnitTest.cs

```diff
@@ -392,6 +417,72 @@ public void DoesNotHandleBackButtonWhenNoNavStack ()
 			Assert.False (result);
 		}
 
+		[Test]
+		public void TestInsertPage()
+		{
+			var root = new ContentPage { Title = "Root" };
+			var newPage = new ContentPage();
+			var navPage = new NavigationPage(root);
+
+			navPage.Navigation.InsertPageBefore(newPage, navPage.RootPage);
+
+			Assert.AreSame(newPage, navPage.RootPage);
+			Assert.AreNotSame(newPage, navPage.CurrentPage);
+			Assert.AreNotSame(navPage.RootPage, navPage.CurrentPage);
+			Assert.AreSame(root, navPage.CurrentPage);
+
+			Assert.Throws<ArgumentException>(() =>
+			{
+				navPage.Navigation.InsertPageBefore(new ContentPage(), new ContentPage());
+			});
+
+			Assert.Throws<ArgumentException>(() =>
+			{
+				navPage.Navigation.InsertPageBefore(navPage.RootPage, navPage.CurrentPage);
+			});
+
+			Assert.Throws<ArgumentException>(() =>
+			{
+				navPage.Navigation.InsertPageBefore(null, navPage.CurrentPage);
+			});
+
+			Assert.Throws<ArgumentException>(() =>
+			{
+				navPage.Navigation.InsertPageBefore(new ContentPage(), null);
+			});
+		}
+
+		[Test]
+		public async void TestRemovePage()
+		{
+			var root = new ContentPage { Title = "Root" };
+			var newPage = new ContentPage();
+			var navPage = new NavigationPage(root);
+			await navPage.PushAsync(newPage);
+
+			navPage.Navigation.RemovePage(root);
+
+			Assert.AreSame(newPage, navPage.RootPage);
+			Assert.AreSame(newPage, navPage.CurrentPage);
+			Assert.AreSame(navPage.RootPage, navPage.CurrentPage);
+			Assert.AreNotSame(root, navPage.CurrentPage);
+
+			Assert.Throws<ArgumentException>(() =>
+			{
+				navPage.Navigation.RemovePage(root);
+			});
+
+			Assert.Throws<InvalidOperationException>(() =>
+			{
+				navPage.Navigation.RemovePage(newPage);
+			});
+
+			Assert.Throws<ArgumentException>(() =>
+			{
+				navPage.Navigation.RemovePage(null);
+			});
+		}
```

Comment:
Surprised there were no test cases for these or I just missed them. This should cover everything (hopefully).